exoscale/api/compute.py

```diff
@@ -2001,6 +2197,121 @@ def get_instance_type(self, name=None, id=None):
+            zone (Zone): the zone in which to create the instance pool
+            name (str): the name of the instance pool
+            description (str): a description of the instance pool
+            size (int): the number of Compute instance members the pool must manage
+            instance_type (InstanceType): the Compute instance members type
+            instance_template (InstanceTemplate): the Compute instance template to use
+                for members
+            instance_root_disk_size (int): the Compute instance members root disk size
+                in GB
+            instance_private_networks ([PrivateNetwork]): a list of Private Networks to
+                attach the Compute instance members to
+            instance_security_groups ([SecurityGroup]): a list of Security Groups to
+                attach the Compute instance members to
+            instance_enable_ipv6 (bool): a flag indicating whether to enable IPv6 on
+                the Compute instance members public network interface
+            instance_ssh_key (SSHKey): a SSH Key to deploy on the Compute instance
+                members
+            instance_user_data (str): a cloud-init user data configuration to apply to
+                the Compute instance members
+
+        Returns:
+            InstancePool: the Compute Instance Pool created
+        """
+
+        if size < 0:
+            raise ValueError("size must be >= 0")
+
+        try:
+            res = self.cs.createInstancePool(
+                name=name,
+                description=description,
+                size=size,
+                zoneid=zone.id,
+                serviceofferingid=instance_type.id,
+                templateid=instance_template.id,
+                rootdisksize=instance_root_disk_size,
+                securitygroupids=list(i.id for i in instance_security_groups)
+                if instance_security_groups
+                else None,
+                networkids=list(i.id for i in instance_private_networks)
+                if instance_private_networks
+                else None,
+                ip6=instance_enable_ipv6,
+                keypair=instance_ssh_key.name if instance_ssh_key else None,
+                userdata=b64encode(bytes(instance_user_data, encoding="utf-8"))
+                if instance_user_data
+                else None,
+            )
+        except CloudStackApiException as e:
+            raise APIException(e.error["errortext"], e.error)
+
+        return InstancePool._from_cs(self, res)
+
+    def list_instance_pools(self, zone, **kwargs):
+        """
+        List Compute Instance Pools.
+
+        Parameters:
+            zone (Zone): a zone to restrict results to
+
+        Yields:
+            InstancePool: the next Compute Instance Pool
+        """
+
+        try:
+            _list = self.cs.listInstancePools(zoneid=zone.id, fetch_list=True, **kwargs)
+
+            for i in _list:
+                yield InstancePool._from_cs(self, i)
```

Comment:
Well originally the `list_*` functions actually returned lists, but in a past code review I've been suggested to use generators instead so here we are :D

Reply:
`yield` is useful in that kind of cases, when there is client-side filtering involved: https://github.com/exoscale/python-exoscale/blob/master/exoscale/api/compute.py#L133-L137

It can be a matter of personal preference, but this is nice to read and understand. Doing the equivalent with a list comprehension would hurt legibility a bit, but it's true that the generator behavior is a bit unfortunate in this case so we could reverse the approach -- that's what the `0.x` series is for.

In this particular case (and perhaps most non-nic-related cases), `return [InstancePool._from_cs(self, i) for i in _list]` would be more developer-friendly.